server/src/main/java/org/opensearch/index/translog/TranslogDeletionPolicy.java

```diff
@@ -165,7 +187,28 @@ synchronized long minTranslogGenRequired(List<TranslogReader> readers, TranslogW
             minByAgeAndSize = Math.max(minByAge, minBySize);
         }
         long minByNumFiles = getMinTranslogGenByTotalFiles(readers, writer, retentionTotalFiles);
-        return Math.min(Math.max(minByAgeAndSize, minByNumFiles), minByLocks);
+        long minByTranslogGenSettings = Math.min(Math.max(minByAgeAndSize, minByNumFiles), minByLocks);
+        return Math.min(minByTranslogGenSettings, minByRetentionLeasesAndSize);
+    }
+
+    static long getMinTranslogGenByRetentionLease(List<TranslogReader> readers, TranslogWriter writer,
+                                                  Supplier<RetentionLeases> retentionLeasesSupplier) {
+        long minGen = writer.getGeneration();
+        final long minimumRetainingSequenceNumber = retentionLeasesSupplier.get()
+            .leases()
+            .stream()
+            .mapToLong(RetentionLease::retainingSequenceNumber)
+            .min()
+            .orElse(Long.MAX_VALUE);
+
+        for (int i = readers.size() - 1; i >= 0; i--) {
+            final TranslogReader reader = readers.get(i);
+            if(reader.getCheckpoint().minSeqNo <= minimumRetainingSequenceNumber &&
+                reader.getCheckpoint().maxSeqNo >= minimumRetainingSequenceNumber) {
+                minGen = Math.min(minGen, reader.getGeneration());
+            }
```

Comment:
Common logic with other methods starting with `getMinTranslog`. Can we abstract out the common code?